File: docs/scripts/Threes.js
import Cell from "./Cell.js";
import SlidingNumbers from "./SlidingNumbers.js";
import EndOfGame from "./EndOfGame.js";

export default class Threes extends SlidingNumbers {
    constructor() {
        super("Threes", "threes_icon.png");
    }

    getRandomValue(cell) {
        let { x, y } = cell;
        this.board[x][y] = this.nextInt(3) + 1;
        this.newNumber.push({ cell: cell, value: this.board[x][y] });
    }

    moveDirection(bRow, bCol, rowDir, colDir) {
        let moved = false, x = bRow + rowDir, y = bCol + colDir;
        if (!this.onBoard(new Cell(x, y))) {
            return moved;
        }
        if (this.board[x][y] !== 0) {
            let sum = this.board[x][y] + this.board[bRow][bCol];
            if (sum === 3 || (sum % 3 === 0 && this.board[x][y] === this.board[bRow][bCol])) {
                moved = this.swap(bRow, bCol, x, y, sum);
                this.score += this.board[x][y];
            }
        } else {
            moved = this.swap(bRow, bCol, x, y);
        }
        return moved;
    }

    canMoveCell(row, col) {
        let positions = [new Cell(row - 1, col), new Cell(row + 1, col), new Cell(row, col - 1), new Cell(row, col + 1)];
        let value = this.board[row][col];
        return positions.some(cell => this.onBoard(cell) && (this.board[cell.getX()][cell.getY()] + value === 3 || ((this.board[cell.getX()][cell.getY()] + value) % 3 === 0) && this.board[cell.getX()][cell.getY()] === value));
    }

    isGameOver() {
        let list = this.listOfCells(0);
        if (list.length === 0 && this.lostGame()) {
            return EndOfGame.LOSE;
        }
        return EndOfGame.NONE;
    }
}

File: src/main/java/model/rules/Threes.java
package model.rules;

import java.util.List;

import model.Cell;
import model.CellValue;
import model.EndOfGame;

public class Threes extends SlidingNumbers {

    protected void getRandomValue(Cell cell) {
        int x = cell.x(), y = cell.y();
        this.board[x][y] = "" + (this.nextInt(3) + 1);
        this.newNumbers.add(new CellValue(cell, this.board[x][y]));
    }

    protected boolean moveDirection(int bRow, int bCol, int rowDir, int colDir) {
        boolean moved = false;
        int x = bRow + rowDir, y = bCol + colDir;
        if (!this.onBoard(new Cell(x, y))) {
            return moved;
        }
        if (this.board[x][y] != "0") {
            int sum = Integer.parseInt(this.board[x][y]) + Integer.parseInt(this.board[bRow][bCol]);
            if (sum == 3 || (sum % 3 == 0 && this.board[x][y].equals(this.board[bRow][bCol]))) {
                moved = this.swap(bRow, bCol, x, y, "" + sum);
                this.score += Integer.parseInt(this.board[x][y]);
            }
        } else {
            moved = this.swap(bRow, bCol, x, y, null);
        }
        return moved;
    }

    protected boolean canMoveCell(int row, int col) {
        var positions = List.of(new Cell(row - 1, col), new Cell(row + 1, col), new Cell(row, col - 1), new Cell(row, col + 1));
        var value = this.board[row][col];
        return positions.stream().anyMatch(
                cell -> this.onBoard(cell) && (Integer.parseInt(this.board[cell.x()][cell.y()]) + Integer.parseInt(value) == 3
                        || ((Integer.parseInt(this.board[cell.x()][cell.y()]) + Integer.parseInt(value)) % 3 == 0)
                                && this.board[cell.x()][cell.y()] == value));
    }

    protected EndOfGame isGameOver() {
        var list = this.listOfCells("0");
        if (list.isEmpty() && this.lostGame()) {
            return EndOfGame.LOSE;
        }
        return EndOfGame.NONE;
    }

}